### ISSUE TITLE
--strict for counting warnings as errors

### ISSUE THREAD

**Issue by ericclemmons** [eslint/eslint#2309]
Soliciting feedback on the idea before doing the work.  My apologies if this is already doable, but I didn't find anything thus far...

### Scenario

- Linting is ran before `nodemon` restarts, the build runs, etc.
- ESLint ensures that any syntax issues cause an error, preventing unnecessary server restarts, builds, etc.
- Warnings are available in the console for the developer to _optionally_ clean up.
- Code _works_, builds _run_, servers _restart_, but there's no way to actually **enforce  standards** without making everything incorrectly an _error_.

**Solution – `--strict`**.

- A `pre-push` git hook can run `eslint --strict` and even _warnings_ count as errors, preventing pushing.
- Continuous integration tools can run in strict mode so that the build breaks when quality tests don't pass.
- The user can still work on a feature with warnings, since the most important thing is the _feature working_.

The gist is, **quality should still be enforced somewhere in the process, but not be an impediment to development.**

Thoughts?

**Comment by michaelficarra** [comment]
I would just make them errors. The author is going to need to fix them before committing, so why not now? The "feature working" argument doesn't convince me; then just don't run the linter.

**Comment by ericclemmons** [comment]
@michaelficarra As indicated in the Scenario above, when you're working on a complex project, you'll likely have `console.log` littered everywhere, several unused vars & things.

The linter also catches syntax & JSX errors, which is **crucial** to know before restarting several services/fixtures between changes.

I don't want these in the final project output, but they're a **necessity** for debugging and getting a feature working.

It's unreasonable & costly to expect complex development to be 100% clean the whole time, which is why **warnings shouldn't impede development, while errors should**.  Similarly, **warnings shouldn't exist in the final, merged code**, IMO.

**Comment by michaelficarra** [comment]
Again, just don't run the linter during development. Your one counterexample (syntax errors) can be solved by running a parser against the code as you write it, something many editors (including vim) will do for you.

**Comment by ericclemmons** [comment]
@michaelficarra I'll wait for someone else to comment.  You're "solutions" (change editors! add more tools! don't run linters!) are unreasonably simplistic and ignorant of several aspects of my use-case.

I _want_ the linter, as there are non-syntax errors that **need** to be caught for an app to function, such as a `<Component />` not having a `displayName` (which is a _requirement_ for some tools).

What I need are answers from the core team on:

- [ ] Is there a way to force _warnings_ as _errors_ with ESLint?
- [ ] If not, is it reasonable to for this option to exist & submit a PR?
- [ ] If not, is there an alternative solution?

**Comment by nzakas** [comment]
I'm not in favor of adding an option to change the error level of rules in this way, I think it will cause more confusion than anything else. I'd advise you to use a different config file in each environment if this is your desired behavior.

**Comment by ericclemmons** [comment]
Ah, feedback from a contributor!  Thanks @nzakas.

I was afraid that was going to be the proposal.  That's unnecessary duplication, IMO, so maybe an alternative solution would be to run the linter and ensure the output is 0 lines?

At least you didn't suggest to change editors :)

**Comment by ericclemmons** [comment]
@nzakas I guess I was looking for something complimentary to:
>   `--quiet`  Report errors only - default: `false`

Where `eslint` would return an exit code of `1` when _anything_ came back.

**Comment by michaelficarra** [comment]
@ericclemmons:

> At least you didn't suggest to change editors :)

I was not telling you to change editors. I was telling you to run a parser against your code during development instead of a linter, something that could be done via CLI, but is also done continuously in many editors.

> Ah, feedback from a contributor!

I've made [the most contributions](https://github.com/eslint/eslint/graphs/contributors) of anyone except @nzakas. But that shouldn't matter.

You are being extremely rude. My "simplistic" answers are trying to pry a better justification out of you. I do this with nearly everyone that opens a poorly-justified "I have an idea" issue on the repos I follow. For ideas with merit, it ends up in a really supportive argument. For weak ideas, they fall over quickly.

**Comment by ericclemmons** [comment]
@michaelficarra Sorry I was rude.  Most responses from open-source contributors are ignorantly assuming the simplistic use-case, rather than what brought this suggestion about in the first place.

So instead, like this one, most of the discussion is having to education others about specific problems out in the wild and is fucking _exhausting_.  Seriously.  Especially when you're working with a problem set that most others probably aren't dealing with.

Instead, I'll pipe `eslint` through `wc` or something and get the results I need.

Thanks!

**Comment by nzakas** [comment]
There's a big difference between filtering messages and changing the exit code. Your proposal blurs the kind between error and warning, and that is problematic.

You can easily create a script to read in the config with warnings and output a new one with errors that is used for your build system. It is duplication, but duplication that is easily automated.

**Comment by ericclemmons** [comment]
I think (the recently landed patch by @rexxars) #2373 solves this issue pretty well.

```js
// .eslint-strict
{
    "extends": ".eslintrc",

    "rules": {
        // Override any settings from the "parent" configuration with stricter requirements
        "no-console": 2
    }
}
```

**Comment by joshkel** [comment]
[`--max-warnings`](http://eslint.org/docs/user-guide/command-line-interface#max-warnings) can serve this purpose: `--max-warnings=0` means that any warnings cause eslint to report an error.

**Comment by ericclemmons** [comment]
@joshkel WHaaaaaaaaaaaaaaaaaaaaaat?

Thanks, I didn't know :)

**Comment by webOS101** [comment]
Just to throw in my $.02, I am in favor of a `--strict` option as well.  `--max-warnings=0` is a nice work-around for use with our CI tool, but it would be nice to have the full list of warnings show up in the CI output.  We already support two flavors of our eslint config, normal and strict.  Having a 3rd, 'ultra-strict', set would be annoying and would likely result in options not getting updated in all places.

The need is that tools like `create-react-app` lint before building, so treating all warnings as errors would prevent building during debugging, as @ericclemmons noted.  In our workflow, warnings are fine during development but should be disallowed during commits.

**Comment by webOS101** [comment]
Well, never mind.  Looks like it still outputs them all. I guess I should have tested it before shooting off my mouth.

**Comment by rhys-vdw** [comment]
Old issue, but if anybody is looking for a solution, this is what we're doing:

```
var path = require('path')

function isTruthy(value) {
  if (!value) return false;
  return ['1', 'true'].indexOf(value.toLowerCase()) >= 0;
}

// Warnings are errors in CI
var OFF = 'off';
var ERROR = 'error';
var WARNING = isTruthy(process.env.CI) ? ERROR : 'warn';


module.exports = {
  // ...
  "rules": {
    "comma-dangle": OFF,
    "eqeqeq": [WARNING, "allow-null"],
    "import/imports-first": OFF,
    "indent": [WARNING, 2, {"SwitchCase": 1}],
    "max-len": [WARNING, 100, 2],
    "no-console": [WARNING, {"allow": ["warn", "error"]}],
    "no-debugger": WARNING,
    "no-fallthrough": WARNING,
    "no-unreachable": WARNING,
    "no-unused-vars": [WARNING, {"vars": "all", "args": "none"}],
    "no-var": ERROR,
    "prefer-const": WARNING,
    "react/prop-types": [WARNING, {"ignore": ["className"]}],
    "semi": [WARNING, "never"],
  },
  // ...
}
```
